### PR TITLE
Cleanup coins from ceased sidechains

### PR DESCRIPTION
**Pull request by abi87**
I have introduced a sidechainHandler to remove coins from ceased sidechains [of course only those coins immediately preceding the cease epoch]. Also forward transfers to ceased sidechains are forbidden.
I have changed the serialization of BlockUndo in order to record key infos for removed coins.
A nice side-effect is that I have centralized and unit tested the sidechain expiration checks

### CODE REVIEW

File: src/coins.cpp [hunk 1]
```diff
@@ -782,6 +829,9 @@ bool CCoinsViewCache::RestoreImmatureBalances(int blockHeight, const CBlockUndo&
 
             cacheSidechains.at(scId).flag = CSidechainsCacheEntry::Flags::DIRTY;
         }
+
+        cacheSidechains.at(scId).scInfo.lastCertificateHash = lastCertHash;
```
Comment (alsala) [inline src/coins.cpp]: for sake of uniformity this should be removed from here and added to 827 above, together with the setting of targetScInfo.lastEpochReferencedByCertificate, that would also spare a useless DIRTY set

File: src/gtest/test_sidechain_ceased.cpp [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+};
+
+
+///////////////////////////////////////////////////////////////////////////////
+/////////////////////////// isSidechainCeased /////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
+TEST_F(CeasedSidechainsTestSuite, UnknownSidechainIsNeitherAliveNorCeased) {
+    uint256 scId = uint256S("aaa");
+    int creationHeight = 1912;
+    CTransaction scCreationTx = txCreationUtils::createNewSidechainTxWith(scId, CAmount(10));
```
Comment (alsala) [inline src/gtest/test_sidechain_ceased.cpp]: Creating this is probably useless

File: src/sc/sidechain.h [hunk 3]
```diff
@@ -33,6 +55,9 @@ class CSidechain {
     // last epoch for which a certificate have been received
     int lastEpochReferencedByCertificate;
```
Comment (alsala) [inline src/sc/sidechain.h:56]: This should also be added in the output of getscinfo RPC cmd

File: src/sc/sidechain.h [hunk 4]
```diff
@@ -78,6 +109,15 @@ namespace Sidechain {
     bool hasScCreationOutput(const CTransaction& tx, const uint256& scId);
 
     bool checkCertSemanticValidity(const CScCertificate& cert, CValidationState& state);
+
+
+    enum class state {
```
Comment (alsala) [inline src/sc/sidechain.h:114]: This also should be added in the output of getscinfo RPC cmd

File: src/gtest/test_sidechain_ceased.cpp (hunk 2 above)
Comment (abi87) [reply src/gtest/test_sidechain_ceased.cpp]: Agree, and it doesn't even help readability. Removed

File: src/sc/sidechain.h (hunk 3 above)
Comment (abi87) [reply src/sc/sidechain.h:56]: Agree, but as discussed AFK,  let's make it subject on subsequent PR

File: src/sc/sidechain.h (hunk 4 above)
Comment (abi87) [reply src/sc/sidechain.h:114]: Agree, but as discussed AFK,  let's make it subject on subsequent PR

**Pull request merged**